// nestjs-rmq/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERROR_TYPE = exports.INITIALIZATION_STEP_DELAY = exports.DEFAULT_PREFETCH_COUNT = exports.DEFAULT_TIMEOUT = exports.DEFAULT_HEARTBEAT_TIME = exports.DEFAULT_RECONNECT_TIME = exports.ERROR_TIMEOUT = exports.ERROR_UNDEFINED_FROM_RPC = exports.ERROR_NO_QUEUE = exports.ERROR_NO_ROUTE = exports.ERROR_NONE_RPC = exports.REPLY_QUEUE = exports.CONNECTED_MESSAGE = exports.DISCONNECT_MESSAGE = exports.CONNECT_EVENT = exports.DISCONNECT_EVENT = exports.RMQ_MODULE_OPTIONS = exports.RMQ_ROUTES_VALIDATE = exports.RMQ_ROUTES_PATH = exports.RMQ_ROUTES_OPTIONS = exports.RMQ_MESSAGE_META = exports.RMQ_ROUTES_META = void 0;
exports.RMQ_ROUTES_META = 'RMQ_ROUTES_META';
exports.RMQ_MESSAGE_META = 'RMQ_MESSAGE_META';
exports.RMQ_ROUTES_OPTIONS = 'RMQ_ROUTES_OPTIONS';
exports.RMQ_ROUTES_PATH = 'RMQ_ROUTES_PATH';
exports.RMQ_ROUTES_VALIDATE = 'RMQ_ROUTES_VALIDATE';
exports.RMQ_MODULE_OPTIONS = 'RMQ_MODULE_OPTIONS';
exports.DISCONNECT_EVENT = 'disconnect';
exports.CONNECT_EVENT = 'connect';
exports.DISCONNECT_MESSAGE = 'Disconnected from RMQ. Trying to reconnect';
exports.CONNECTED_MESSAGE = 'Successfully connected to RMQ';
exports.REPLY_QUEUE = 'amq.rabbitmq.reply-to';
exports.ERROR_NONE_RPC = 'This is none RPC queue. Use notify() method instead';
exports.ERROR_NO_ROUTE = 'Requested service doesn\'t have RMQRoute with this path';
exports.ERROR_NO_QUEUE = 'No queueName specified! You will not recieve messages in RMQRoute';
exports.ERROR_UNDEFINED_FROM_RPC = 'RPC method returned undefined';
exports.ERROR_TIMEOUT = 'Response timeout error';
exports.DEFAULT_RECONNECT_TIME = 5;
exports.DEFAULT_HEARTBEAT_TIME = 5;
exports.DEFAULT_TIMEOUT = 30000;
exports.DEFAULT_PREFETCH_COUNT = 0;
exports.INITIALIZATION_STEP_DELAY = 300;
var ERROR_TYPE;
(function (ERROR_TYPE) {
    ERROR_TYPE["TRANSPORT"] = "TRANSPORT";
    ERROR_TYPE["RMQ"] = "RMQ";
})(ERROR_TYPE = exports.ERROR_TYPE || (exports.ERROR_TYPE = {}));
