// nestjs-rmq/dist/rmq.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var RMQModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RMQModule = void 0;
const rmq_service_1 = require("./rmq.service");
const common_1 = require("@nestjs/common");
const rmq_metadata_accessor_1 = require("./rmq-metadata.accessor");
const rmq_explorer_1 = require("./rmq.explorer");
const core_1 = require("@nestjs/core");
const constants_1 = require("./constants");
const rmq_error_service_1 = require("./rmq-error.service");
let RMQModule = RMQModule_1 = class RMQModule {
    static forRoot(options) {
        return {
            module: RMQModule_1,
            providers: [rmq_service_1.RMQService, { provide: constants_1.RMQ_MODULE_OPTIONS, useValue: options }],
            exports: [rmq_service_1.RMQService],
        };
    }
    static forRootAsync(options) {
        const asyncOptions = this.createAsyncOptionsProvider(options);
        return {
            module: RMQModule_1,
            imports: options.imports,
            providers: [rmq_service_1.RMQService, rmq_metadata_accessor_1.RMQMetadataAccessor, rmq_explorer_1.RMQExplorer, asyncOptions],
            exports: [rmq_service_1.RMQService],
        };
    }
    static createAsyncOptionsProvider(options) {
        return {
            provide: constants_1.RMQ_MODULE_OPTIONS,
            useFactory: (...args) => __awaiter(this, void 0, void 0, function* () {
                const config = yield options.useFactory(...args);
                return config;
            }),
            inject: options.inject || [],
        };
    }
};
RMQModule = RMQModule_1 = __decorate([
    common_1.Global(),
    common_1.Module({
        imports: [core_1.DiscoveryModule],
        providers: [rmq_metadata_accessor_1.RMQMetadataAccessor, rmq_explorer_1.RMQExplorer, rmq_error_service_1.RmqErrorService]
    })
], RMQModule);
exports.RMQModule = RMQModule;
