// nestjs-rmq/dist/classes/rmq-extended-message.class.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtendedMessage = void 0;
class ExtendedMessage {
    constructor(msg) {
        this.content = msg.content;
        this.fields = msg.fields;
        this.properties = msg.properties;
    }
    getDebugString() {
        try {
            const content = JSON.parse(this.content.toString());
            const debugMsg = {
                fields: this.fields,
                properties: this.properties,
                message: this.maskBuffers(content),
            };
            return JSON.stringify(debugMsg);
        }
        catch (e) {
            return e.message;
        }
    }
    maskBuffers(obj) {
        let result = {};
        for (let prop in obj) {
            if (obj[prop].type === 'Buffer') {
                result[prop] = 'Buffer - length ' + obj[prop].data.length;
            }
            else {
                result[prop] = obj[prop];
            }
        }
        return result;
    }
}
exports.ExtendedMessage = ExtendedMessage;
