// nestjs-rmq/dist/helpers/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RQMColorLogger = void 0;
const common_1 = require("@nestjs/common");
const chalk_1 = require("chalk");
class RQMColorLogger {
    constructor(logMessages) {
        this.logMessages = logMessages !== null && logMessages !== void 0 ? logMessages : false;
    }
    log(message, context) {
        common_1.Logger.log(message, context);
    }
    error(message, trace, context) {
        common_1.Logger.error(message, trace, context);
    }
    debug(message, context) {
        if (!this.logMessages) {
            return;
        }
        const split = new RegExp(/(.*?)(\[.*?])(.*)/g).exec(message);
        if (split[3]) {
            common_1.Logger.log(`${chalk_1.blueBright(split[1])} ${chalk_1.yellow(split[2])} ${chalk_1.white(split[3])}`);
        }
        else {
            common_1.Logger.log(message, context);
        }
    }
    warn(message, context) {
        common_1.Logger.warn(message, context);
    }
}
exports.RQMColorLogger = RQMColorLogger;
